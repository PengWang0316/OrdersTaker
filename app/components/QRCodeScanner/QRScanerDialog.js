import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Dialog, DialogTitle, Typography, Button } from '@material-ui/core';
import { withStyles } from '@material-ui/core/styles';
import QrReader from 'react-qr-reader';

import { setTableNumber as setTableNumberAction } from '../../actions/CartActions';

let handleClose;
let handleSetTableNumber;

const styles = {
  buttonDiv: {
    display: 'flex',
    justifyContent: 'flex-end'
  }
};

/**
 * Calling the redux action and close the dialog.
 * @param {object} data contains value that reads from QR code
 * @return {null} No return.
 */
const handleScan = data => {
  if (data && (/^\d+$/.test(data) || typeof data === 'number')) {
    handleSetTableNumber(data);
    handleClose();
  }
};

const handleError = err => console.error(err);

export const QRScanerDialog = ({
  onClose, isOpen, classes, setTableNumber
}) => {
  handleClose = onClose;
  handleSetTableNumber = setTableNumber;
  return (
    <Dialog onClose={onClose} open={isOpen}>
      <DialogTitle><Typography color="primary">Please scan the QR code on the table</Typography></DialogTitle>
      <QrReader onScan={handleScan} onError={handleError} />
      <div className={classes.buttonDiv}>
        <Button size="small" onClick={onClose}>Cancel</Button>
      </div>
    </Dialog>
  );
};
QRScanerDialog.propTypes = {
  classes: PropTypes.object.isRequired,
  onClose: PropTypes.func.isRequired,
  isOpen: PropTypes.bool.isRequired,
  setTableNumber: PropTypes.func.isRequired
};
/* istanbul ignore next */
const mapDispatchToProps = dispatch => ({
  setTableNumber: number => dispatch(setTableNumberAction(number))
});
export default connect(null, mapDispatchToProps)(withStyles(styles)(QRScanerDialog));
