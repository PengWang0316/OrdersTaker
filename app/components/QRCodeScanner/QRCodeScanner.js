import React, { Component, Fragment } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { PhotoCamera as PhotoCameraIcon } from '@material-ui/icons';
import { Button, Typography } from '@material-ui/core';
import { withStyles } from '@material-ui/core/styles';

import QRScanerDialog from './QRScanerDialog';

const styles = {
  tableNumber: {
    fontSize: 35,
    fontWeight: '800'
  }
};

/**
 * The component that will be used to scan the QR code.
 */
export class QRCodeScanner extends Component {
  static propTypes = {
    cart: PropTypes.object.isRequired,
    classes: PropTypes.object.isRequired
  };

  state = {
    isDialogOpen: false
  };

  /**
   * Toggle the dialog open state
   * @return {null} No return
   */
  toggleDialog = () => this.setState(({ isDialogOpen }) => ({ isDialogOpen: !isDialogOpen }));

  /**
 * The render method
 * @return {jsx} Return jsx for the component.
 */
  render() {
    const { cart, classes } = this.props;
    const { isDialogOpen } = this.state;
    return (
      <div>
        {!cart.tableNumber && (
          <Fragment>
            <Button onClick={this.toggleDialog}><PhotoCameraIcon color="primary" /></Button>
            <Typography color="primary">Scan Table QR</Typography>
          </Fragment>
        )}
        {cart.tableNumber && (
          <Fragment>
            <Typography color="primary">Table number:</Typography>
            <Button color="primary" className={classes.tableNumber} onClick={this.toggleDialog}>{cart.tableNumber}</Button>
          </Fragment>
        )}
        <QRScanerDialog isOpen={isDialogOpen} onClose={this.toggleDialog} />
      </div>
    );
  }
}
/* istanbul ignore next */
const mapStateToProps = state => ({
  cart: state.cart
});
export default connect(mapStateToProps, null)(withStyles(styles)(QRCodeScanner));
