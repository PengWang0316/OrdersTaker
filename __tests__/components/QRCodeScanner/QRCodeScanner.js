import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import { QRCodeScanner } from '../../../app/components/QRCodeScanner/QRCodeScanner';

jest.mock('@material-ui/core/Button', () => 'Button');
jest.mock('@material-ui/core/Typography', () => 'Typography');
jest.mock('@material-ui/icons/PhotoCamera', () => 'PhotoCamera');
jest.mock('../../../app/components/QRCodeScanner/QRScanerDialog', () => 'QRScanerDialog');

describe('QRCodeScanner', () => {
  const defaultProps = {
    classes: {
      tableNumber: 'tableNumber'
    },
    orders: {}
  };
  const getShallowComponent = (props = defaultProps) => shallow(<QRCodeScanner {...props} />);

  test('Initial state and toggleDialog function', () => {
    const component = getShallowComponent();
    expect(component.state('isDialogOpen')).toBe(false);
    component.instance().toggleDialog();
    expect(component.state('isDialogOpen')).toBe(true);
  });

  test('Snapshot without tableNumber', () => expect(renderer.create(<QRCodeScanner {...defaultProps} />).toJSON()).toMatchSnapshot());
  test('Snapshot with tableNumber', () => expect(renderer.create(<QRCodeScanner {...{ ...defaultProps, orders: { tableNumber: 1 } }} />).toJSON()).toMatchSnapshot());
});
