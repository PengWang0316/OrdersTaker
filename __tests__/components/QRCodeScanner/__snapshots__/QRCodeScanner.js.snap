// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QRCodeScanner Snapshot with tableNumber 1`] = `
<div>
  <Typography
    color="primary"
  >
    Table number:
  </Typography>
  <Button
    className="tableNumber"
    color="primary"
    onClick={[Function]}
  >
    1
  </Button>
  <QRScanerDialog
    isOpen={false}
    onClose={[Function]}
  />
</div>
`;

exports[`QRCodeScanner Snapshot without tableNumber 1`] = `
<div>
  <Button
    onClick={[Function]}
  >
    <PhotoCamera
      color="primary"
    />
  </Button>
  <Typography
    color="primary"
  >
    Scan Table QR
  </Typography>
  <QRScanerDialog
    isOpen={false}
    onClose={[Function]}
  />
</div>
`;
